src/wp-includes/global-styles-and-settings.php

```diff
@@ -83,18 +83,34 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
  * @return string Stylesheet.
  */
 function wp_get_global_stylesheet( $types = array() ) {
-	// Return cached value if it can be used and exists.
-	// It's cached by theme to make sure that theme switching clears the cache.
-	$can_use_cached = (
-		( empty( $types ) ) &&
-		( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) &&
-		( ! defined( 'SCRIPT_DEBUG' ) || ! SCRIPT_DEBUG ) &&
-		( ! defined( 'REST_REQUEST' ) || ! REST_REQUEST ) &&
-		! is_admin()
-	);
-	$transient_name = 'global_styles_' . get_stylesheet();
+	/*
+	 *
+	 * Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme
+	 * developer's workflow.
+	 *
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	$can_use_cached = empty( $types ) && ! WP_DEBUG;
+	/*
+	 * By using the 'theme_json' group, this data is marked to be non-persistent across requests.
+	 * See `wp_cache_add_non_persistent_groups` in src/wp-includes/load.php and other places.
+	 *
+	 * The rationale for this is to make sure derived data from theme.json
+	 * is always fresh from the potential modifications done via hooks
+	 * that can use dynamic data (modify the stylesheet depending on some option,
+	 * settings depending on user permissions, etc.).
+	 * See some of the existing hooks to modify theme.json behaviour:
+	 * https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/
+	 *
+	 * A different alternative considered was to invalidate the cache upon certain
+	 * events such as options add/update/delete, user meta, etc.
+	 * It was judged not enough, hence this approach.
+	 * See https://github.com/WordPress/gutenberg/pull/45372
+	 */
+	$cache_group    = 'theme_json';
+	$cache_key      = 'wp_get_global_stylesheet';
 	if ( $can_use_cached ) {
-		$cached = get_transient( $transient_name );
+		$cached = wp_cache_get( $cache_key, $cache_group );
```

Comment:
I'm concerned that we may be reintroducing the bug from https://core.trac.wordpress.org/ticket/56975#comment:32 here. Have you tested that this problem isn't caused by adding calls to `wp_cache_*()` methods here? We need to make sure the function isn't called anywhere when `load-styles.php` is used.

Reply:
Thanks @felixarntz for the reminder to test this PR for the same issue that occurred with `load-styles.php` on the other PR.

I did test it. [See my test report here.](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1405464211) No fatal errors and `load-styles.php` loaded properly ✅ 

And to make sure, I added `wp_cache_get()` back into `wp_theme_has_theme_json()` and reran the tests. Tests failed as the admin area broke.

So in my testing, this PR does not affect `load-styles.php`.

Would be good to have another person test ran the same tests just to make sure. cc @ironprogrammer